Add simple ODE demo

jwallwork23 opened the pull request:
Closes #100.

Also introduces a new `PointSeq` class for use with ODE problems.

jwallwork23 commented:
Test suite seems to be failing for the same reason currently under investigation in #104.

jwallwork23 commented:
@ddundo Since you mentioned you'd already been looking at this, would you be happy to review it?

ddundo reviewed (approved):
Looks nice @jwallwork23! Left a few small comments :)

jwallwork23 commented on goalie/point_seq.py:
```diff
@@ -0,0 +1,46 @@
+        :kwarg get_initial_condition: a function, whose only argument is a
+            :class:`~.MeshSeq`, which specifies initial conditions on the first mesh
+        :kwarg get_form: a function, whose only argument is a :class:`~.MeshSeq`, which
+            returns a function that generates the PDE weak form
```
Good spot - this was a copy-paste error. Addressed in e4a9708.

jwallwork23 commented on demos/ode.py:
```diff
@@ -0,0 +1,323 @@
+solutions = point_seq.solve_forward()["u"]["forward"]
+forward_euler_trajectory = [1] + [
+    sol.dat.data[0] for by_sub in solutions for sol in by_sub
+]
```
Good points! Addressed in 3e0b6a4e24ee144cdc582328287b4120dec2ab0d and 32a804984fb4aace428c1d6967be02e50aeb4c93.

jwallwork23 commented on demos/ode.py:
```diff
@@ -0,0 +1,323 @@
+# .. math::
+#    \frac{\mathrm{d}u}{\mathrm{d}t} = u,\quad u(0) = 1,
+#
+# which we solve for :math:`u(t)` over the time period :math:`t\in[0,1]`. It is
```
Fixed in 002900e.

jwallwork23 commented on demos/ode.py:
```diff
@@ -0,0 +1,323 @@
+# the timestep length. This expression may be rearranged as
+#
+# .. math::
+#    u_{i+1} - u_i + - \Delta t u_i = 0.
```
Good spot. Fixed in 323c0c3.

jwallwork23 commented on demos/ode.py:
```diff
@@ -0,0 +1,323 @@
+# function and integrate:
+#
+# .. math::
+#    \int (u_{i+1} - u_i  - \Delta t u_i) v \mathrm{d}R = 0, \forall v\in R.
```
Fixed in aaadddd, also adding integration bounds.

jwallwork23 commented on demos/ode.py:
```diff
@@ -0,0 +1,323 @@
+# Similarly to the above, this gives rise to
+#
+# .. math::
+#    \int (u_{i+1} - u_{i} - \Delta t u_{i+1}) v \mathrm{d}R, \forall v\in R.
```
Fixed in aaadddd, also adding integration bounds.

jwallwork23 commented on demos/ode.py:
```diff
@@ -0,0 +1,323 @@
+
+backward_euler_trajectory = [1] + [
+    sol.dat.data[0] for by_sub in solutions for sol in by_sub
+]
```
Addressed in 3e0b6a4e24ee144cdc582328287b4120dec2ab0d.

jwallwork23 commented:
Thanks for the review @ddundo. I think I addressed all your concerns. Also used `float(sol)` instead of `sol.dat.data[0]` for simplicity and to avoid exposing underlying Firedrake data structures in an early demo.

ddundo commented:
Thanks @jwallwork23, I think this is much clearer now!
